Unable to generate coverage report using lcov on MacOs 15.3.1

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current behaviour

While trying to get a coverage report , the report generation fails complaining of duplicate function definitions.

### Expected behaviour

Generate a coverage report for the branch.

### Steps to reproduce

```
cmake -B build -DCMAKE_BUILD_TYPE=Coverage
cmake --build build
cmake -P build/Coverage.cmake
```

### Relevant log output

```
Capturing coverage data from src
geninfo cmd: '/opt/homebrew/Cellar/lcov/2.3/bin/geninfo src --toolname lcov --output-filename baseline.info --initial --gcov-tool /Users/prabhatverma/projects/bitcoin/build/cov_tool_wrapper.sh'
Found LLVM gcov version 16.0.0, which emulates gcov version 4.2.0
Using intermediate gcov format
Recording 'internal' directories:
        /Users/prabhatverma/projects/bitcoin/build/src
        src
Writing temporary data to /tmp/geninfo_datI48l
Scanning src for .gcno files ...
Found 401 graph files in src
using: chunkSize: 1, nchunks:401, intervalLength:20
lcov: WARNING: (gcov) GCOV did not produce any data for /Users/prabhatverma/projects/bitcoin/build/src/secp256k1/src/CMakeFiles/secp256k1_precomputed.dir/precomputed_ecmult.c.gcno
        (use "lcov --ignore-errors gcov,gcov ..." to suppress this warning)
lcov: WARNING: (gcov) GCOV did not produce any data for /Users/prabhatverma/projects/bitcoin/build/src/secp256k1/src/CMakeFiles/secp256k1_precomputed.dir/precomputed_ecmult_gen.c.gcno
Message summary:
  1 error message:
    inconsistent: 1
  2 warning messages:
    gcov: 2
lcov: ERROR: (inconsistent) "/Users/prabhatverma/projects/bitcoin/src/secp256k1/src/scalar_impl.h":67: duplicate function 'secp256k1_scalar_split_lambda' starts on line 67 but previous definition started on 142 while capturing from src/secp256k1/src/CMakeFiles/exhaustive_tests.dir/tests_exhaustive.c.gcno.
        (use "lcov --ignore-errors inconsistent ..." to bypass this error)
CMake Error at build/CoverageInclude.cmake:45 (execute_process):
  execute_process failed command indexes:

    1: "Child return code: 1"

Call Stack (most recent call first):
  build/Coverage.cmake:5 (include)
```

### How did you obtain Bitcoin Core

Compiled from source

### What version of Bitcoin Core are you using?

master@5b8fd7c

### Operating system and version

MacOs Sequoia 15.3.1

### Machine specifications

_No response_

--- a/doc/developer-notes.md
+++ b/doc/developer-notes.md
@@ -484,6 +484,8 @@ $ ./build/test/functional/test_runner.py --valgrind
 
 ### Compiling for test coverage
 
+#### Using LCOV
+
 LCOV can be used to generate a test coverage report based upon `ctest`
 execution. LCOV must be installed on your system (e.g. the `lcov` package
 on Debian/Ubuntu).
@@ -512,6 +514,63 @@ To enable test parallelism:
 ```
 cmake -DJOBS=$(nproc) -P build/Coverage.cmake
 ```
+
+#### Using LLVM/Clang toolchain
+
+The following generates a coverage report for unit tests and functional tests.
+
+Configure the build with the following flags:
+
+> Consider building with a clean state using `rm -rf build`
+
+```shell
+# MacOS may instead require `-DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++"`
+cmake -B build -DCMAKE_C_COMPILER="clang" \
+   -DCMAKE_CXX_COMPILER="clang++" \
+   -DAPPEND_CFLAGS="-fprofile-instr-generate -fcoverage-mapping" \
+   -DAPPEND_CXXFLAGS="-fprofile-instr-generate -fcoverage-mapping" \
+   -DAPPEND_LDFLAGS="-fprofile-instr-generate -fcoverage-mapping"
+cmake --build build # Use "-j N" here for N parallel jobs.
+```
+
+Generating the raw profile data based on `ctest` and functional tests execution:
+
+```shell
+# Create directory for raw profile data
+mkdir -p build/raw_profile_data
+
+# Run tests to generate profiles
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" build/test/functional/test_runner.py # Use "-j N" here for N parallel jobs
+
+# Merge all the raw profile data into a single file
+find build/raw_profile_data -name "*.profraw" | xargs llvm-profdata merge -o build/coverage.profdata
+```
+
+> **Note:** The "counter mismatch" warning can be safely ignored, though it can be resolved by updating to Clang 19.
+> The warning occurs due to version mismatches but doesn't affect the coverage report generation.
+
+Generating the coverage report:
+
+```shell
+llvm-cov show \
+    --object=build/bin/test_bitcoin \
+    --object=build/bin/bitcoind \
+    -Xdemangler=llvm-cxxfilt \
+    --instr-profile=build/coverage.profdata \
+    --ignore-filename-regex="src/crc32c/|src/leveldb/|src/minisketch/|src/secp256k1/|src/test/" \
+    --format=html \
+    --show-instantiation-summary \
+    --show-line-counts-or-regions \
+    --show-expansions \
+    --output-dir=build/coverage_report \
+    --project-title="Bitcoin Core Coverage Report"
+```
+
+> **Note:** The "functions have mismatched data" warning can be safely ignored, the coverage report will still be generated correctly despite this warning.
+> This warning occurs due to profdata mismatch created during the merge process for shared libraries.
+
+The generated coverage report can be accessed at `build/coverage_report/index.html`.
 
 ### Performance profiling with perf
 

